app/script/event/EventRepository.js

```diff
@@ -698,6 +698,44 @@ z.event.EventRepository = class EventRepository {
     return this.conversationService.update_event(identifiedUpdates);
   }
 
+  _handleDuplicatedEvent(originalEvent, newEvent) {
+    switch (newEvent.type) {
+      case z.event.Client.CONVERSATION.MESSAGE_ADD:
+        return this._handleLinkPreviewUpdate(originalEvent, newEvent);
+
+      case z.event.Client.CONVERSATION.ASSET_ADD:
+        return this._handleAssetUpdate(originalEvent, newEvent);
+
+      default:
+        this._throwValidationError(newEvent, `Forbidden type '${newEvent.type}' for duplicate events`);
+    }
+  }
+
+  _handleAssetUpdate(originalEvent, newEvent) {
+    const newData = newEvent.data;
+    switch (newData.status) {
+      case z.assets.AssetTransferState.UPLOADED: {
+        const updatedData = Object.assign({}, originalEvent.data, newData);
+        const updatedEvent = Object.assign({}, originalEvent, {data: updatedData});
+        return this.conversationService.update_event(updatedEvent);
+      }
+
+      case z.assets.AssetTransferState.UPLOAD_FAILED: {
+        // case of both failed or canceled upload
+        const fromSelf = newEvent.from === this.userRepository.self().id;
+        const uploadFailed = newEvent.data.reason === z.assets.AssetUploadFailedReason.FAILED;
+        if (fromSelf && uploadFailed) {
+          // we only want to keep the message if it was sent by the user and the upload failed for some reason
+          return this.conversationService.update_asset_as_failed_in_db(originalEvent.primary_key, newEvent.data.reason);
+        }
+        // in all the other cases (cancelation or receiving failed upload) we just delete the event
+        return this.conversationService.delete_message_from_db(newEvent.conversation, newEvent.id).then(() => newEvent);
+      }
+      default:
+        return this._throwValidationError(newEvent, `Unhandled asset status update '${newEvent.data.status}'`);
+    }
+  }
```

Comment:
There are asset events that do not contain and `status`. `original` and `prievew` containing a thumbnail for videos or a loudness graph for audio files will be thrown out by this but should not.